File: components/dashboard/progress/weight-chart-utils.ts
import type { TimeRange } from '@/lib/core/types/dashboard';

export function buildXTicks(
  count: number,
  range: TimeRange,
  locale: string,
  nowLabel: string,
  weekPrefix: string
): { ticks: number[]; formatter: (v: number, idx: number) => string } {
  const uniqueTicks = (values: number[]) =>
    values.filter((value, index, array) => array.indexOf(value) === index);

  if (count < 2) {
    return {
      ticks: [0],
      formatter: () => nowLabel,
    };
  }

  if (range === '30d') {
    const step = Math.floor(count / 4);
    const ticks = uniqueTicks([0, step, step * 2, step * 3, count - 1]);
    return {
      ticks,
      formatter: (_v: number, idx: number) =>
        idx === ticks.length - 1 ? nowLabel : `${weekPrefix}${idx + 1}`,
    };
  }

  // 90d — show months
  const today = new Date();
  const monthFormatter = new Intl.DateTimeFormat(locale, { month: 'short' });
  const ticks: number[] = [];
  const labels: string[] = [];
  for (let i = 2; i > 0; i--) {
    const d = new Date(today);
    d.setMonth(d.getMonth() - i);
    const dayIndex = count - 1 - i * 30;
    ticks.push(Math.max(0, dayIndex));
    labels.push(monthFormatter.format(d));
  }
  ticks.push(count - 1);
  labels.push(nowLabel);
  return {
    ticks: uniqueTicks(ticks),
    formatter: (_v: number, idx: number) => labels[idx] ?? '',
  };
}

File: apps/mobile-flutter/lib/features/dashboard/logic/weight_chart_utils.dart
/// Vendored VERBATIM from the RN `lib/dashboard/logic/weight-chart-utils.ts` —
/// keep in sync.
///
/// Provides the X-axis ticks + a label formatter: W1/W2/…/Now for 30d,
/// month-short/Now for 90d.
library;

import 'package:intl/intl.dart';

import '../../../models/dashboard.dart';

/// Result of [buildXTicks]: the tick indices plus a label formatter that takes
/// `(value, idx)` and returns the displayed string.
class XTicks {
  const XTicks(this.ticks, this.formatter);

  final List<int> ticks;
  final String Function(int value, int idx) formatter;
}

List<int> _uniqueTicks(List<int> values) {
  final seen = <int>{};
  final out = <int>[];
  for (final v in values) {
    if (seen.add(v)) out.add(v);
  }
  return out;
}

/// Builds X-axis ticks + formatter for the weight chart. [count] is the number
/// of data points; [range] selects the labeling strategy.
XTicks buildXTicks(
  int count,
  WeightRange range,
  String locale,
  String nowLabel,
  String weekPrefix,
) {
  if (count < 2) {
    return XTicks([0], (_, __) => nowLabel);
  }

  if (range == WeightRange.d30) {
    final step = count ~/ 4;
    final ticks = _uniqueTicks([0, step, step * 2, step * 3, count - 1]);
    return XTicks(
      ticks,
      (_, idx) => idx == ticks.length - 1 ? nowLabel : '$weekPrefix${idx + 1}',
    );
  }

  // 90d — show months.
  final today = DateTime.now();
  final monthFormatter = DateFormat.MMM(locale);
  final ticks = <int>[];
  final labels = <String>[];
  for (var i = 2; i > 0; i--) {
    final d = DateTime(today.year, today.month - i, today.day);
    final dayIndex = count - 1 - i * 30;
    ticks.add(dayIndex < 0 ? 0 : dayIndex);
    labels.add(monthFormatter.format(d));
  }
  ticks.add(count - 1);
  labels.add(nowLabel);
  return XTicks(
    _uniqueTicks(ticks),
    (_, idx) => idx < labels.length ? labels[idx] : '',
  );
}
